perf(ngcc): spawn workers eagerly

alan-agius4 opened the pull request:
With this change we spawn workers eagerly based on the amount of work that needs to be done and changed the behaviour in `analyzeEntryPoints`, to only create tasks for non-processed formats.

Before this change we spawned the maximum of workers possible. However, in some cases there are less tasks than the max number of workers which resulted in created unnecessary workers

Reference: #35717

gkalpak commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -207,6 +207,10 @@ export function mainNgcc(
       for (const formatProperty of propertiesToProcess) {
+        if (hasBeenProcessed(entryPoint.packageJson, formatProperty)) {
+          continue;
+        }
```
Suggested changes:
- Remove the `hasBeenProcessed` check from the `CompileFn` below.
- Retain the current logging behavior: ``logger.debug(`Skipping ${entryPoint.name} : ${formatProperty} (already compiled).`);``
- Remove [TaskProcessingOutcome.AlreadyProcessed][1], as it is not used any more. (Theoretically, we could remove `TaskProcessingOutcome` altogether, but I would keep it in case we need to add other outcomes in the future (e.g. handle errors differently).)

[1]: https://github.com/angular/angular/blob/d3c0d92d2/packages/compiler-cli/ngcc/src/execution/api.ts#L78-L79

gkalpak commented on packages/compiler-cli/ngcc/src/main.ts:
```diff
@@ -33,7 +33,7 @@ import {ParallelTaskQueue} from './execution/task_selection/parallel_task_queue'
 import {ConsoleLogger, LogLevel} from './logging/console_logger';
 import {Logger} from './logging/logger';
-import {hasBeenProcessed, markAsProcessed} from './packages/build_marker';
+import {hasBeenProcessed} from './packages/build_marker';
```
💯

gkalpak reviewed (changes requested):
Thx, @alan-agius4 :tada:

Can you please split this up into two commits (one for lazily workers and one for checking `hasBeenProcessed` upfront)?
Also, in the commit message: `eagerly` --> `lazily` :grin:

alan-agius4 committed:
perf(ngcc): spawn workers lazily

With this change we spawn workers lazily based on the amount of work that needs to be done.

Before this change we spawned the maximum of workers possible. However, in some cases there are less tasks than the max number of workers which resulted in created unnecessary workers

Reference: #35717

alan-agius4 committed:
perf(ngcc): only create tasks for non-processed formats

Change the behaviour in `analyzeEntryPoints` to only create tasks for non-processed formats.

gkalpak commented on packages/compiler-cli/ngcc/src/execution/cluster/master.ts:
```diff
@@ -256,11 +256,11 @@ export class ClusterMaster {
    * Wrap an event handler to ensure that `finishedDeferred` will be rejected on error (regardless
    * if the handler completes synchronously or asynchronously).
    */
-  private wrapEventHandler<Args extends unknown[]>(fn: (...args: Args) => void|Promise<void>):
```
Why this change? Are there performance (or other) implications?
If not, I would prefer to keep it, because otherwise it is easy to introduce an error by making a handler async.

alan-agius4 replied:
I don't think that is correct, `cluster.on` is not a callback based methods using `async/await` in it might result in unexpected behaviour. Though probably, in this case might not, since you are not interested in the resolution of the `wrapEventHandler` itself.

alan-agius4 replied:
Or am I missing something?

alan-agius4 replied:
Actually I take that back!

gkalpak replied:
So, what is your current opinion on it, I'm confused 😕 :grin:

alan-agius4 replied:
I reverted the commit

gkalpak reviewed (approved):
Nice! Thx, @alan-agius4 :+1:

AndrewKushnir commented:
@gkalpak @alan-agius4 quick question: should this change be master-only (i.e. go to `9.1.0`) or it should also be included in `9.0.x`? Currently it's merged to master only (according to the `PR target` label).

alan-agius4 commented:
@AndrewKushnir, I checked and it should apply cleanly. So let's merge it to the branch path as well.

atscott commented:
Hi @alan-agius4, since this was already merged and closed, can you open a new PR that targets patch only and copy these changes in?

alan-agius4 commented:
@atscott here you go https://github.com/angular/angular/pull/35832
